sys/kern/src/task.rs

```diff
@@ -820,47 +820,55 @@ pub fn check_task_id_against_table(
 /// Selects a new task to run after `previous`. Tries to be fair, kind of.
 ///
 /// If no tasks are runnable, the kernel panics.
-pub fn select(previous: usize, tasks: &[Task]) -> usize {
-    priority_scan(previous, tasks, |t| t.is_runnable())
-        .expect("no tasks runnable")
+pub fn select(previous: usize, tasks: &[Task]) -> &Task {
+    match priority_scan(previous, tasks, |t| t.is_runnable()) {
+        Some((_index, task)) => task,
+        None => panic!(),
+    }
 }
 
+/// Scans the task table to find a prioritized candidate.
+///
 /// Scans `tasks` for the next task, after `previous`, that satisfies `pred`. If
 /// more than one task satisfies `pred`, returns the most important one. If
 /// multiple tasks with the same priority satisfy `pred`, prefers the first one
-/// in order after `previous`, mod `tasks.len()`.
+/// in order after `previous`, mod `tasks.len()`. Finally, if no tasks satisfy
+/// `pred`, returns `None`
 ///
 /// Whew.
 ///
 /// This is generally the right way to search a task table, and is used to
 /// implement (among other bits) the scheduler.
 ///
-/// # Panics
-///
-/// If `previous` is not a valid index in `tasks`.
+/// On success, the return value is the task's index in the task table, and a
+/// direct reference to the task.
 pub fn priority_scan(
     previous: usize,
     tasks: &[Task],
     pred: impl Fn(&Task) -> bool,
-) -> Option<usize> {
-    uassert!(previous < tasks.len());
-    let search_order = (previous + 1..tasks.len()).chain(0..previous + 1);
-    let mut choice = None;
-    for i in search_order {
-        if !pred(&tasks[i]) {
+) -> Option<(usize, &Task)> {
```

Comment:
From a scan of the use sites, it looks like most want a scheduling hint (`NextTask`) as their return value. I hesitate to add a lifetime parameter to `NextTask`, as it's quite commonly used and would extend a lot of borrows in ways that might prove awkward.

The intent here would be to eliminate possible bounds checks in the consumption of the `NextTask` value it currently returns? I'd want to double-check that there are bounds checks being generated today before adding the complexity, but it might be an improvement.

Reply:
I was indeed thinking of eliminating bounds checks.

I'll see about checking this option out (at a lower priority), and if it proves to be an easy change then I'll check the asm output to see if there is any potential benefit there.

Reply:
This probably also makes no difference, or is negative. The `unsafe extern "C" fn handle_fault` function includes a ton of `handle_fault` calls but already does not seemingly have a single bounds check panic site. Trying to remove from that would obviously only cause an unsigned integer overflow panic.